retroshare-service/src/retroshare-service.cc

```diff
@@ -17,66 +17,166 @@
+                 std::cout <<'*';
+             else
+                 std::cout << ch,std::cout.flush();
+         }
+    }
+  std::cout <<std::endl;
+  return password;
+}
+
+class RsServiceNotify: public NotifyClient
+{
+public:
+	RsServiceNotify(){}
+	virtual ~RsServiceNotify() {}
+
+	virtual bool askForPassword(const std::string& title, const std::string& question, bool prev_is_bad, std::string& password,bool& cancel)
+	{
+		std::string question1=title + "\nPlease enter your PGP password for key:\n    " + question + " :";
+		password = readStringFromKeyboard(question1.c_str()) ;
+		cancel = false ;
+
+		return !password.empty();
+	}
+};
+
+
 int main(int argc, char* argv[])
 {
 #ifdef __ANDROID__
 	AndroidStdIOCatcher dbg; (void) dbg;
 	QAndroidService app(argc, argv);
-#else // def __ANDROID__
-	QCoreApplication app(argc, argv);
-#endif // def __ANDROID__
 
 	signal(SIGINT,   QCoreApplication::exit);
 	signal(SIGTERM,  QCoreApplication::exit);
 #ifdef SIGBREAK
 	signal(SIGBREAK, QCoreApplication::exit);
 #endif // ifdef SIGBREAK
 
+#endif // def __ANDROID__
+
 	RsInit::InitRsConfig();
+	RsControl::earlyInitNotificationSystem();
 
-	// clumsy way to enable JSON API by default
-	if(!QCoreApplication::arguments().contains("--jsonApiPort"))
-	{
-		int argc2 = argc + 2;
-		char* argv2[argc2]; for (int i = 0; i < argc; ++i ) argv2[i] = argv[i];
-		char opt[] = "--jsonApiPort";
-		char val[] = "9092";
-		argv2[argc] = opt;
-		argv2[argc+1] = val;
-		RsInit::InitRetroShare(argc2, argv2, true);
-	}
-	else RsInit::InitRetroShare(argc, argv, true);
+#ifndef __ANDROID__
+	RsServiceNotify *notify = new RsServiceNotify();
+	rsNotify->registerNotifyClient(notify);
+#endif
 
-	RsControl::earlyInitNotificationSystem();
+#ifndef __ANDROID__
+        std::string webui_pass1 = "Y";
+        std::string webui_pass2 = "N";
+
+        for(;;)
+        {
+			webui_pass1 = readStringFromKeyboard("Please register a password for the web interface: ");
+			webui_pass2 = readStringFromKeyboard("Please enter the same password again            : ");
+
+            if(webui_pass1 != webui_pass2)
+            {
+                std::cerr << "Passwords do not match!" << std::endl;
+                continue;
+            }
+            if(webui_pass1.empty())
+            {
+                std::cerr << "Password cannot be empty!" << std::endl;
+                continue;
+            }
+
+			break;
+        }
+#endif
+
+    if(RsInit::InitRetroShare(argc, argv, true))
+    {
+        std::cerr << "Could not properly init Retroshare core." << std::endl;
+        return 1;
+    }
+
+#ifdef __ANDROID__
 	rsControl->setShutdownCallback(QCoreApplication::exit);
+
 	QObject::connect(
 	            &app, &QCoreApplication::aboutToQuit,
 	            [](){
 		if(RsControl::instance()->isReady())
 			RsControl::instance()->rsGlobalShutDown(); } );
 
 	return app.exec();
+#else
+    if(jsonApiServer)
+		jsonApiServer->authorizeToken("webui:"+webui_pass1);
+
+	while(true)
```

Comment:
How do you get out of this? You need `rsControl->setShutdownCallback` also on other systems not only in android, you can use an `std::atomic<bool> keepRunning` captured by reference in the lambada passed as callback that get set to false in the callback

Reply:
this is still pending. I'll follow your advice.

Reply:
done